cora/signal/lss.py

```diff
@@ -875,6 +997,328 @@ def za_density_grid(
+
+        Returns
+        -------
+        smoothed_field
+            The field with the smoothing applied.
+        """
+
+        # Just return the input if FoGs are effectively turned off
+        if self.alpha_FoG == 0.0:
+            return field
+
+        # Get the growth factor and smoothing scales
+        D = field.cosmology.growth_factor(field.redshift)
+        sigmaP = self._sigmaP(field.redshift)
+
+        K = exponential_FoG_kernel(field.chi, self.alpha_FoG * sigmaP, D)
+
+        smoothed_field = BiasedLSS(axes_from=field, attrs_from=field)
+
+        # Apply the smoothing kernel and directly assign into output field
+        np.matmul(K, field.delta[:], out=smoothed_field.delta[:])
+
+        return smoothed_field
+
+    def _sigmaP(self, z: FloatArrayLike) -> FloatArrayLike:
+        """Get the velocity dispersion at a given redshift.
+
+        Parameters
+        ----------
+        z
+            The redshift to calculate at.
+
+        Returns
+        -------
+        sigmav
+            The velocity dispersion in Mpc/h.
+        """
+
+        # Evalulate the polynomial bias model to get the default bias
+        return np.sum(
+            [c * (z - self.z_eff) ** n for n, c in enumerate(self.FoG_coeff)], axis=0
+        )
+
+
+class AddCorrelatedShotNoise(RandomTask, task.SingleTask):
+    """Add a correlated shot noise contribution to each input map.
+
+    The shot noise realisation is deterministically generated from the LSS field, so all
+    tasks receiving the same `InitialLSS` object as a requirement will generate the same
+    shot noise realisation.
+
+    Attributes
+    ----------
+    n_eff : float
+        The effective number density in sources per Mpc^3.
+    log_M_HI_g : float
+        Set the shot noise level by specifying the log10 of the average HI mass per
+        tracer galaxy. This is overriden if `n_eff` is set explicitly.
+    omega_HI_model : str
+        The model to use for Omega_HI when applying the mean 21cm temperature. See
+        :func:`omega_HI` for the list of valid models and details. Default is
+        `Crighton2015`.
+    """
+
+    n_eff = config.Property(proptype=float, default=None)
+    log_M_HI_g = config.Property(proptype=float, default=None)
+    omega_HI_model = config.enum(_omega_HI_models, default="Crighton2015")
+
+    def setup(self, lss: InitialLSS):
+        """Set up a common seed from the LSS field.
+
+        Parameters
+        ----------
+        lss
+            Common InitialLSS field to use for seeding.
+        """
+        import zlib
+
+        # Get a subset of the data in bytes
+        lss_subset = lss.delta[:, :100].copy().tobytes()
+
+        # If the seed was not set, hash the input LSS and use that
+        if self.seed is None:
+            self.seed = zlib.adler32(lss_subset)
+
+        if self.n_eff is not None:
+            self._n_eff_z = np.ones_like(lss.chi) * self.n_eff
+        elif self.log_M_HI_g is not None:
```

Comment:
The M_HI-to-n_eff conversion is wrong. Here's what I came up with as a replacement:
```
h = lss.cosmology.H0 / 100
H0_SI = corr.cosmology.H(0) # H_0 in s^-1
omHI = omega_HI(lss.redshift, model=self.omega_HI_model)
M_HI_g = (10 ** log10MHI) * units.solar_mass # in kg

# First compute M_HI / rho_HI(z) in SI units (m^3)
self.n_eff_z = 8 * np.pi * units.G * M_HI_g / (3.0 * omHI * H0_SI ** 2)
# Then convert to h^-3 Mpc^3
self.n_eff_z *= h ** 3 / units.mega_parsec ** 3
# Finally take inverse, which yields the equivalent n_eff for the input M_HI
self.n_eff_z = 1 / self.n_eff_z
```
This matches my own calculation (using `halomod`, not attempting to use precisely the same cosmological parameters etc.) to 8%, which is good enough.

Reply:
Well that's embarrassing. Totally wrong. I guess I shouldn't code late at night. Thanks for finding the issues.